ZMQ: Create "sequence" notifier, enabling client-side mempool tracking

instagibbs opened the pull request:
This PR creates a new ZMQ notifier that gives a "total hash history" of block (dis)connection, mempool addition/substraction, all in one pipeline. It also exposes a "mempool sequence number" to both this notifier and `getrawmempool` results, which allows the consumer to use the results together without confusion about ordering of results and without excessive `getrawmempool` polling.

See the functional test `interfaces_zmq.py::test_mempool_sync` which shows the proposed user flow for the client-side tracking of mempool contents and confirmations.

Inspired by https://github.com/bitcoin/bitcoin/pull/19462#issuecomment-656140421
Alternative to https://github.com/bitcoin/bitcoin/pull/19462 due to noted deficiencies in current zmq notification streams.

Also fixes a legacy zmq test that didn't actually trigger a reorg because of identical blocks being generated on each side of the split(oops)

instagibbs committed:
zmq test: Actually make reorg occur

instagibbs commented:
Getting some hopefully unrelated test errors for `mempool_packages.py`. https://travis-ci.org/github/bitcoin/bitcoin/jobs/711137623

promag commented on test/functional/interface_zmq.py:
```diff
@@ -147,8 +147,8 @@ def test_reorg(self):
-        # Generate 2 blocks in nodes[1]
-        self.nodes[1].generatetoaddress(2, ADDRESS_BCRT1_UNSPENDABLE)
+        # Generate 2 blocks in nodes[1] to a different address to ensure split
+        self.nodes[1].generatetoaddress(2, ADDRESS_BCRT1_P2WSH_OP_TRUE)
```
951507c4d1152ceab2a930fb16fdac05504f0056

oops 😅 good catch.

promag commented on src/txmempool.h:
```diff
@@ -725,6 +731,17 @@ class CTxMemPool
     }
 
+    /** Guards this internal counter for external reporting */
+    uint32_t GetAndIncrementSequence() const {
```
`EXCLUSIVE_LOCKS_REQUIRED(cs)` instead?

instagibbs replied:
done

promag reviewed (commented):
I still think #19476 makes life easier for the client in regards to fault tolerance and simplicity. With this PR, the client still has to bake its state from RPC and ZMQ.

I think you could extract `CValidationInterface` changes to a different PR.

instagibbs commented:
From conversations with people who use the notifications already they
seemed more keen to a solution like this that doesn't require long polling
support. Also it makes wumpus less upset to just improve a current
notification system 🤯

That's said they're not necessarily in competition.

On Thu, Jul 23, 2020, 6:37 PM João Barbosa <notifications@github.com> wrote:

> *@promag* commented on this pull request.
>
> I still think #19476 <https://github.com/bitcoin/bitcoin/pull/19476>
> makes life easier for the client in regards to fault tolerance and
> simplicity. With this PR, the client still has to bake its state from RPC
> and ZMQ.
>
> I think you could extract CValidationInterface changes to a different PR.
> ------------------------------
>
> In test/functional/interface_zmq.py
> <https://github.com/bitcoin/bitcoin/pull/19572#discussion_r459761022>:
>
> > @@ -147,8 +147,8 @@ def test_reorg(self):
>
>          self.nodes[0].generatetoaddress(1, ADDRESS_BCRT1_UNSPENDABLE)
>
>          assert_equal(self.nodes[0].getbestblockhash(), hashblock.receive().hex())
>
>
>
> -        # Generate 2 blocks in nodes[1]
>
> -        self.nodes[1].generatetoaddress(2, ADDRESS_BCRT1_UNSPENDABLE)
>
> +        # Generate 2 blocks in nodes[1] to a different address to ensure split
>
> +        self.nodes[1].generatetoaddress(2, ADDRESS_BCRT1_P2WSH_OP_TRUE)
>
>
> 951507c
> <https://github.com/bitcoin/bitcoin/commit/951507c4d1152ceab2a930fb16fdac05504f0056>
>
> oops 😅 good catch.
> ------------------------------
>
> In src/txmempool.h
> <https://github.com/bitcoin/bitcoin/pull/19572#discussion_r459763375>:
>
> > @@ -725,6 +731,17 @@ class CTxMemPool
>
>          return (m_unbroadcast_txids.count(txid) != 0);
>
>      }
>
>
>
> +    /** Guards this internal counter for external reporting */
>
> +    uint32_t GetAndIncrementSequence() const {
>
>
> EXCLUSIVE_LOCKS_REQUIRED(cs) instead?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/19572#pullrequestreview-454535546>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABMAFU6BN2AS2HWUGITFY7LR5C3UFANCNFSM4PF2HZKQ>
> .
>

promag commented:
> That's said they're not necessarily in competition.

I wasn't suggesting otherwise 👍

jonasschnelli commented:
Concept ACK

instagibbs commented:
I'm running some extensive testing of this PR against testnet and mainnet nodes, tracking expected mempool for a couple weeks using a modified test harness. Once I'm satisfied I'll be updating the test for this PR with the updated logic(if any) and maybe the consumer tool itself.

instagibbs commented:
added test showing conflicted tx from block announcement("R" shows up before "C")

instagibbs commented:
Was able to replicate a seeming gap in mempool_sequence values being reported. Pushed a possibly controversial fix(touching mempool logic) showing that it fixes the test case written.

While that is a *very* minimal change I can understand it might rub people the wrong way.

The alternatives are:
0) Ignore the gaps, it doesn't seem to meaningfully interact with using the features as-is. Consumer will be tracking zmq sequence value, and any `getrawmempool` results will report an "expected" value.
1) "Just" account for these gaps. When hitting an "R" mempool_sequence gap, sum up those gaps, then make sure that value + the number of found mempool transactions being removed normally equals the mempool_sequence reported in the next mempool report.
2) Add another removal reporting hook which just announces all txhashes being removed.

(2) seems like way overkill. (0) is fine for mempool view tracking, (1) just allows the consumer to do more asserting I guess, you can take it or leave it.

Thoughts on last commit?

luke-jr commented on src/init.cpp:
```diff
@@ -483,10 +483,12 @@ void SetupServerArgs(NodeContext& node)
     gArgs.AddArg("-zmqpubhashtx=<address>", "Enable publish hash transaction in <address>", ArgsManager::ALLOW_ANY, OptionsCategory::ZMQ);
     gArgs.AddArg("-zmqpubrawblock=<address>", "Enable publish raw block in <address>", ArgsManager::ALLOW_ANY, OptionsCategory::ZMQ);
     gArgs.AddArg("-zmqpubrawtx=<address>", "Enable publish raw transaction in <address>", ArgsManager::ALLOW_ANY, OptionsCategory::ZMQ);
+    gArgs.AddArg("-zmqpubsequence=<address>", "Enable publish hash block and tx sequence in <address>", ArgsManager::ALLOW_ANY, OptionsCategory::ZMQ);
```
"sequence" doesn't really express what this does. Maybe "zmqpubmempool"?

instagibbs commented:
Point taken but it also publishes a sequence of block hashes for connects
and disconnects.

On Thu, Jul 30, 2020, 8:32 PM Luke Dashjr <notifications@github.com> wrote:

> *@luke-jr* commented on this pull request.
> ------------------------------
>
> In src/init.cpp
> <https://github.com/bitcoin/bitcoin/pull/19572#discussion_r463343776>:
>
> > @@ -483,10 +483,12 @@ void SetupServerArgs(NodeContext& node)
>      gArgs.AddArg("-zmqpubhashtx=<address>", "Enable publish hash transaction in <address>", ArgsManager::ALLOW_ANY, OptionsCategory::ZMQ);
>      gArgs.AddArg("-zmqpubrawblock=<address>", "Enable publish raw block in <address>", ArgsManager::ALLOW_ANY, OptionsCategory::ZMQ);
>      gArgs.AddArg("-zmqpubrawtx=<address>", "Enable publish raw transaction in <address>", ArgsManager::ALLOW_ANY, OptionsCategory::ZMQ);
> +    gArgs.AddArg("-zmqpubsequence=<address>", "Enable publish hash block and tx sequence in <address>", ArgsManager::ALLOW_ANY, OptionsCategory::ZMQ);
>
> "sequence" doesn't really express what this does. Maybe "zmqpubmempool"?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/19572#pullrequestreview-458848709>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABMAFU4SHWAGULPMYAO77NLR6IGKNANCNFSM4PF2HZKQ>
> .
>

instagibbs commented:
rebased on master.

Improved mempool sync test with code I've been running on mainnet/testnet for a few weeks now, last week with no issues. Added exclusive locks annotation.

luke-jr commented on src/validationinterface.h:
```diff
@@ -98,6 +98,17 @@ class CValidationInterface {
+     *
+     * Required because of legacy ZMQ notifications including block
+     * transactions when using TransactionAddedToMempool. See:
+     * CZMQNotificationInterface::TransactionAddedToMempool
```
Ugly hack. Just change CZMQNotificationInterface

instagibbs replied:
Sorry, change it how specifically? If you mean I should change the legacy notifications too, I disagree? Might be misunderstanding.

instagibbs replied:
Figured out what you meant. Done in follow-on commit.

instagibbs commented on src/txmempool.h:
```diff
@@ -474,6 +474,12 @@ class CTxMemPool
+    // In-memory counter for external mempool tracking purposes.
+    // This number is incremented once every time a transaction
+    // is added or removed from the mempool(except when removed
+    // for block inclusion).
```
Right, stale comment. Will update.

instagibbs replied:
done

instagibbs commented on src/txmempool.h:
```diff
@@ -474,6 +474,12 @@ class CTxMemPool
+    // This number is incremented once every time a transaction
+    // is added or removed from the mempool(except when removed
+    // for block inclusion).
+    mutable uint32_t sequence_number{1};
```
I believe my original theory was that the zmq sequence values themselves were also 32 bit, so I figured not much to be gained since any consumer has to handle this wrap around anyways?

instagibbs replied:
Offline someone noted that the ZMQ sequence numbers(probably) reset if the connection itself is reset, but the mempool sequence numbers will continue. I think this is a good argument to expand to `uint64_t`. Will do that soon.

instagibbs replied:
done

instagibbs commented on src/txmempool.h:
```diff
@@ -474,6 +474,12 @@ class CTxMemPool
+    // This number is incremented once every time a transaction
+    // is added or removed from the mempool(except when removed
+    // for block inclusion).
+    mutable uint32_t sequence_number{1};
```
done

instagibbs commented on src/rpc/blockchain.cpp:
```diff
@@ -483,7 +483,7 @@ static void entryToJSON(const CTxMemPool& pool, UniValue& info, const CTxMemPool
-UniValue MempoolToJSON(const CTxMemPool& pool, bool verbose)
+UniValue MempoolToJSON(const CTxMemPool& pool, bool verbose, bool include_mempool_sequence)
 {
     if (verbose) {
```
done, throws error + test

instagibbs commented on src/rpc/blockchain.cpp:
```diff
@@ -499,14 +499,25 @@ UniValue MempoolToJSON(const CTxMemPool& pool, bool verbose)
         }
         return o;
     } else {
+        uint32_t mempool_sequence = 0;
```
done

instagibbs commented on src/init.cpp:
```diff
@@ -499,6 +501,7 @@ void SetupServerArgs(NodeContext& node)
     hidden_args.emplace_back("-zmqpubhashtxhwm=<n>");
```
done

instagibbs commented on src/init.cpp:
```diff
@@ -499,6 +501,7 @@ void SetupServerArgs(NodeContext& node)
     hidden_args.emplace_back("-zmqpubhashtxhwm=<n>");
     hidden_args.emplace_back("-zmqpubrawblockhwm=<n>");
     hidden_args.emplace_back("-zmqpubrawtxhwm=<n>");
+    hidden_args.emplace_back("-zmqpubseqhwm=<n>");
```
done

instagibbs commented:
Updated PR with fixes/cleanups, and changed sequence number to 8 bytes.

Dogfooding consumer still running strong after 2+ weeks, no mempool inconsistencies or dropped messages.

jonatack commented:
Concept ACK

(You had me at "dogfooding")

instagibbs commented:
Looks like the consumer logic failed during the recent mainnet reorg of one block https://twitter.com/BitMEXResearch/status/1296431817592057857

`A->D->A` occurred and the mempool sequence number jumped by a couple hundred between the two `A`s which is unexpected based on my assumptions of Core locking.

instagibbs commented:
After some consideration of the failure detected above, I'm going to stick with the current format, and eat the cost of a single `getrawmempool` call every time an unexpected mempool sequence value is encountered. If all evictions are announced for block inclusion we get dangerously close to blowing out zmq buffers, and in practice is creating more traffic for little gain.

In other words, please continue review :)

instagibbs commented:
rebased

n-thumann commented:
Tested ACK, works as expected on macOS 10.15.6.
Just two minor hints: I think this should also be added to `doc/zmq.md` and explicitly explain the difference between the already existing sequence number and this new mempool sequence number to prevent confusion. Maybe also add these new notifications to `contrib/zmq_sub.py` with unpacking of message type (Tx addition/removal, Block connect/disconnect) and mempool sequence number ✌️

instagibbs commented:
@n-thumann good points, I was unfamiliar with those docs. Will update with an additional commit

instagibbs committed:
Update zmq notification documentation and sample consumer

instagibbs commented:
@n-thumann can you actually get `contrib/zmq/zmq_sub.py` to work? I updated what I think it should look like, but I had trouble running it even with legacy notifications. Try it out?

n-thumann commented on doc/zmq.md:
```diff
@@ -119,3 +121,8 @@ There are several possibilities that ZMQ notification can get lost
 using. Bitcoind appends an up-counting sequence number to each
 notification which allows listeners to detect lost notifications.
+
+The "sequence" publisher refers specifically to the mempool sequence
```
slightly more consistent wording 😊 
```suggestion
The `sequence` topic refers specifically to the mempool sequence
```

n-thumann commented on contrib/zmq/zmq_sub.py:
```diff
@@ -69,6 +68,12 @@ async def handle(self) :
             print('- RAW TX ('+sequence+') -')
             print(binascii.hexlify(body))
+        elif topic == b"sequence":
+            hash = body[:32].hex()
```
Use bytes instead of string
```suggestion
            hash = binascii.hexlify(body[:32])
```

n-thumann commented on contrib/zmq/zmq_sub.py:
```diff
@@ -69,6 +68,12 @@ async def handle(self) :
+            label = chr(body[32])
+            mempool_sequence = None if len(body) != 32+1+8 else struct.unpack("<Q", body[32+1:])[0]
+            print('- SEQUENCE ('+sequence+') -')
+            print("{} {} {}".format(hash, label, mempool_sequence))
```
Format looks unnecessary here :)
```suggestion
            print(hash, label, mempool_sequence)
```

n-thumann commented:
> @n-thumann can you actually get `contrib/zmq/zmq_sub.py` to work?

Your updated script runs perfectly fine!
Another doc related idea: Should [these lines](https://github.com/bitcoin/bitcoin/blame/6bc7b33008b7974f8866a38cc858822923ce5a22/doc/zmq.md#L114-L118) be changed? This only applies when using `hashblock` or `rawblock` and can be circumvented by using `sequence` topic.
I think it´s also useful to briefly mention [here](https://github.com/bitcoin/bitcoin/blame/6bc7b33008b7974f8866a38cc858822923ce5a22/doc/zmq.md#L91) that the body of `sequence` notifications differs from the others, so basically what you´ve already written [here](https://github.com/bitcoin/bitcoin/pull/19572/files#diff-61fb4c079cf5f7c216f97714baa7b224R228-R234) ✌️.
Oh, and in case you have a little bit of spare time I would you mind having a look at my (also zmq-related) PR https://github.com/bitcoin/bitcoin/pull/18309? 😊

instagibbs commented:
took your suggestions @n-thumann , added some more information to docs as well as suggested

jonatack reviewed (commented):
Checkpoint review feedback so far:
- the debug build is clean and tests green
- the contrib script seems to work well and without issues after a day of doing it's thing, though I needed to chmod it to change permissions to run it (EDIT: fixed in #19870)
- the new c++ code could definitely do with some clang formatting :heart: 
- now or for later, there is a fair amount of duplicate-looking code, nothing blocking tho
- so far the test `interface_zmq.py` fails in 38fbf2d8f5 at line 103
- in 9b5daf74f the test fails as well; best if each commit has passing tests
- EDIT: the test passes for me in commits d353bd19b, a976c97c9, and 92ca0fca0

A few minor comments follow to take or leave.

instagibbs commented on src/zmq/zmqnotificationinterface.cpp:
```diff
@@ -179,17 +198,73 @@ void CZMQNotificationInterface::TransactionAddedToMempool(const CTransactionRef&
+            CZMQAbstractNotifier *notifier = *i;
+            if (notifier->NotifyTransaction(*ptx))
+            {
+                i++;
```
is this in the style guide? :P

jonatack replied:
yeah, `++i` is preferred over `i++`

instagibbs replied:
done

jonatack commented on doc/zmq.md:
```diff
@@ -109,13 +119,19 @@ No authentication or authorization is done on connecting clients; it
+It is up to the subscriber to retrieve the chain from the last known
+block to the new tip. Also note that no notification occurs if the tip
+was in the active chain - this is the case after calling invalidateblock RPC.
+In contrast the `sequence` topic publishes all block connects and disconnects.
```
nit proposals, feel free to take or ignore
```diff
-Note that for `*block` topics when the block chain tip changes,
-a reorganisation may occur and just the tip will be notified.
+Note that for `*block` topics, when the block chain tip changes,
+a reorganisation may occur and only the tip will be notified.
 It is up to the subscriber to retrieve the chain from the last known
-block to the new tip. Also note that no notification occurs if the tip
-was in the active chain - this is the case after calling invalidateblock RPC.
-In contrast the `sequence` topic publishes all block connects and disconnects.
+block to the new tip. Also note that no notification will occur if the tip
+was in the active chain--as would be the case after calling the invalidateblock
+RPC. In contrast, the `sequence` topic publishes all block connections and
+disconnections.
```

instagibbs replied:
took most of this

jonatack commented on doc/zmq.md:
```diff
@@ -109,13 +119,19 @@ No authentication or authorization is done on connecting clients; it
+The `sequence` topic refers specifically to the mempool sequence
+number which is also published along with all mempool events. This
+is a different sequence value than in ZMQ itself to allow a total
+ordering of mempool events to be constructed.
```
nit suggestions
```diff
 The `sequence` topic refers specifically to the mempool sequence
-number which is also published along with all mempool events. This
-is a different sequence value than in ZMQ itself to allow a total
+number, which is also published along with all mempool events. This
+is a different sequence value than in ZMQ itself in order to allow a total
 ordering of mempool events to be constructed.
```

instagibbs replied:
done

jonatack reviewed (commented):
Approach ACK on this useful feature for bitcoind mempool management that has the bonus of you and others dogfooding it. Thanks for the tests and updated contrib script and doc. It seems pretty close apart from the nits mentioned above and clean organisation of the commits wrt the test changes. I'm not sure if it's helpful for 92ca0fca04cdb6 to be a separate commit, as it partially undoes some of the changes in the previous commits, unless you wanted to document the design in some way. Otherwise LGTM, will re-review after the next update.

jonatack commented on src/rpc/blockchain.cpp:
```diff
@@ -483,9 +483,12 @@ static void entryToJSON(const CTxMemPool& pool, UniValue& info, const CTxMemPool
 {
     if (verbose) {
+        if (include_mempool_sequence) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Verbose results can not contain mempool sequence values.");
```
s/can not/cannot/

instagibbs replied:
done

instagibbs replied:
my string change broke the test of this. See!?!

jonatack replied:
Proof of test! 😄

jonatack commented on src/rpc/blockchain.cpp:
```diff
@@ -499,14 +502,25 @@ UniValue MempoolToJSON(const CTxMemPool& pool, bool verbose)
+        } else {
+            UniValue o(UniValue::VOBJ);
+            o.pushKV("txids", a);
+            o.pushKV("mempool_sequence", mempool_sequence);
```
The behavior of `getrawmempool` with the additional param and its interaction with `verbose` should be covered in a functional test, unless I missed the added assertion somewhere.

instagibbs replied:
There is a test line for it:
`assert_raises_rpc_error(-8, "Verbose results can not contain mempool sequence values.", self.nodes[0].getrawmempool, True, True)`

jonatack replied:
Oops, I even tested it manually as well. More like an `assert("mempool_sequence" in getrawmempool.keys())` or even checking the value too.

jonatack replied:
Never mind, it looks like you covered that.

jonatack reviewed (commented):
A couple items with the RPC changes.

instagibbs committed:
Add functional tests for zmq sequence topic and mempool sequence logic

instagibbs committed:
Add 'sequence' zmq publisher to track all block (dis)connects, mempool deltas

Using the zmq notifications to avoid excessive mempool polling can be difficult
given the current notifications available. It announces all transactions
being added to mempool or included in blocks, but announces no evictions
and gives no indication if the transaction is in the mempool or a block.

Block notifications for zmq are also substandard, in that it only announces
block tips, while all block transactions are still announced.

This commit adds a unified stream which can be used to closely track mempool:

1) getrawmempool to fill out mempool knowledge
2) if txhash is announced, add or remove from set
based on add/remove flag
3) if blockhash is announced, get block txn list,
remove from those transactions local view of mempool
4) if we drop a sequence number, go to (1)

The mempool sequence number starts at the value 1, and
increments each time a transaction enters the mempool,
or is evicted from the mempool for any reason, including
block inclusion. The mempool sequence number is published
via ZMQ for any transaction-related notification.

These features allow for ZMQ/RPC consumer to track mempool
state in a more exacting way, without unnecesarily polling
getrawmempool. See interface_zmq.py::test_mempool_sync for
example usage.

instagibbs commented on test/functional/interface_zmq.py:
```diff
@@ -204,5 +223,284 @@ def test_reorg(self):
+        #    remaining txs(5) + 1 rbf(A+R) + 1 block connect + 1 final tx
+        expected_sequence = get_raw_seq
+        r_gap = 0
+        for i in range(num_txs+2+1+1):
```
done

instagibbs commented on test/functional/interface_zmq.py:
```diff
@@ -204,5 +223,284 @@ def test_reorg(self):
+
+        # Things continue to happen in the "interim" while waiting for snapshot results
+        # We have node 0 do all these to avoid p2p races with RBF announcements
+        for i in range(num_txs):
```
done

instagibbs commented on test/functional/interface_zmq.py:
```diff
@@ -204,5 +223,284 @@ def test_reorg(self):
+        # or we lost a ZMQ message somehow and want to start over
+        txids = []
+        num_txs = 5
+        for i in range(num_txs):
```
done

instagibbs commented on test/functional/interface_zmq.py:
```diff
@@ -204,5 +223,284 @@ def test_reorg(self):
+
+            # More to be simply mined
+            more_tx = []
+            for i in range(5):
```
done

instagibbs commented on test/functional/interface_zmq.py:
```diff
@@ -5,10 +5,11 @@
-from test_framework.messages import CTransaction, hash256
-from test_framework.util import assert_equal, connect_nodes
+from test_framework.messages import CTransaction, hash256, FromHex
+from test_framework.util import assert_equal, connect_nodes, assert_raises_rpc_error
```
done

instagibbs commented on contrib/zmq/zmq_sub.py:
```diff
@@ -47,16 +48,14 @@ def __init__(self):
-          msgSequence = struct.unpack('<I', msg[-1])[-1]
-          sequence = str(msgSequence)
+        if len(seq) == 4:
+          sequence = str(struct.unpack('<I', seq)[-1])
```
done

instagibbs commented on test/functional/interface_zmq.py:
```diff
@@ -204,5 +223,284 @@ def test_reorg(self):
+        <32-byte hash>A<8-byte LE uint> : Transactionhash added mempool
+        """
+        self.log.info("Testing 'sequence' publisher")
+        import zmq
```
done

instagibbs replied:
found out why: zmq-less functional test builds. Stuck it inside a pass-through try/except block.

instagibbs commented:
@jonatack I'm keeping https://github.com/bitcoin/bitcoin/pull/19572/commits/003767e7a92e8ae41f5c561a69d7be23e4753f63 in for now out of a real reason(it's changing how the legacy zmq notifications work under the hood) and laziness.

Should have addessed everything. Did some history cleanup as well.

instagibbs commented:
@mruddy in case you're interested(I don't know if you're on IRC)

instagibbs commented:
rebased, looks like appveyor and one of the cirrus builds is having the sads

MarcoFalke commented:
You might have to add a suppression, not sure if the tsan failue is a false positive

instagibbs commented:
Can't seem to replicate locally with tsan.

instagibbs commented:
@n-thumann (can't seem to re-request with your stale review, thanks github)

n-thumann commented:
> @n-thumann (can't seem to re-request with your stale review, thanks github)

re-ACK c2f03fd7b9090ac17299190f118511326483f23c ✌️

jonatack reviewed (commented):
A couple of errors while building 712f214.

instagibbs commented:
collapsed the two logic commits since it's not that much code and lots of lines changed twice in a row

jonatack commented on src/rpc/blockchain.cpp:
```diff
@@ -529,6 +544,15 @@ static UniValue getrawmempool(const JSONRPCRequest& request)
+                        RPCResult::Type::OBJ, "", "",
+                        {
+                            {RPCResult::Type::NUM, "mempool_sequence", "The mempool sequence value."},
+                            {RPCResult::Type::ARR, "txids", "",
```
885212e could you please inverse the order of `txids` and `mempool_sequence` fields here in the help to be the same as the rpc output

```diff
                         {
-                            {RPCResult::Type::NUM, "mempool_sequence", "The mempool sequence value."},
                             {RPCResult::Type::ARR, "txids", "",
                             {
                                 {RPCResult::Type::STR_HEX, "", "The transaction id"},
                             }},
+                            {RPCResult::Type::NUM, "mempool_sequence", "The mempool sequence value."},
                         }},
```

instagibbs replied:
done

jonatack commented on test/functional/interface_zmq.py:
```diff
@@ -5,13 +5,24 @@
 
-from test_framework.address import ADDRESS_BCRT1_UNSPENDABLE
+from test_framework.address import ADDRESS_BCRT1_UNSPENDABLE, ADDRESS_BCRT1_P2WSH_OP_TRUE
+from test_framework.blocktools import create_block, create_coinbase, add_witness_commitment
```
99a94fe
```suggestion
from test_framework.blocktools import add_witness_commitment, create_block, create_coinbase
```

jonatack commented on test/functional/interface_zmq.py:
```diff
@@ -5,13 +5,24 @@
+from test_framework.util import (
+    assert_equal,
+    connect_nodes,
+    assert_raises_rpc_error,
```
99a94fe nit sort

jonatack reviewed (commented):
Tested ACK 4e1cc1d88889784e962e4393091c445b6561df9d

Aligning the rpc output and help would be good. Feel free to ignore the test import order nits.

0xB10C reviewed (commented):
Concept ACK for using a ZMQ feed for tracking mempool and chain changes. The sequence number allows to order and detect missing mempool changes. This seems helpful. 

This change allows Bitcoin Core users to keep track of their mempool in an external service connected via ZMQ. It makes the external detection of reorgs easier as every block connection is notified (not only changes to the block tip as its the case with the other `*block` topics).


On the approach: IMO the `sequence` notifier could use proper ZMQ multipart messages instead of in-lining a new protocol. https://github.com/bitcoin/bitcoin/pull/19572/files#diff-d8e6fe13399f13c42a93ec8326c60614R94-R97

```
now:
| topic | <32-byte hash>R<8-byte LE uint> | zmq sequence num |

multipart:
| topic | 32-byte-hash | char | (optional) mempool sequence | zmq sequence | 
```

I mentioned something like this in https://github.com/bitcoin/bitcoin/pull/17878#discussion_r363282493 as a concept alternative as well. However, I think adding proper ZMQ multipart message support to `CZMQAbstractPublishNotifier::SendMessage` could be a change in its own. This could, for example, look like [this SendMessage](https://github.com/0xB10C/bitcoin/commit/a06f5c7db6c0e491f150fe5dafb864c15326bb71#diff-979f11c86130c660e4c2272fc8be1ceaR191-R225) with [this zmq_send_multipart](https://github.com/0xB10C/bitcoin/commit/a06f5c7db6c0e491f150fe5dafb864c15326bb71#diff-979f11c86130c660e4c2272fc8be1ceaR63-R91). Let me know what you think.

0xB10C commented:
Had a quick look at the _Bitcoin Core is restarted and transactions are added to the mempool from `mempool.dat`_ case. These are correctly notified, but don't necessary have the same `mempool_sequence` (that's expected). With a `mempool.dat` containing multiple thousand transactions the ZMQ buffer could fill up when using the `DEFAULT_ZMQ_SNDHWM` of 1000 resulting in ZMQ messages being dropped. This is, however, nothing new and can happen with the existing `*tx` ZMQ publishers as well. 

Here, the ZMQ buffers could fill up when multiple thousand transactions are removed from the mempool (this excludes confirmations). This would result in some removal notifications not reaching the external service. But I believe these events to be rare. It's more something a user of the new ZMQ `sequence` endpoint should keep in mind. Not really specific to this PR.

instagibbs commented:
@0xB10C thanks for the review!

> Discussion about dropped messages
Yes, this is a general problem that can be mostly avoided, and if not the endpoint logic should basically toss what it thinks it knows and restart its tracking.

> Multipart messages 
I'm not sure if the change alone is worth it for current messages. I think it really shines when there are optional messages, of which we currently have none. Is this a fair assessment? A potential application would be for example we publish optional mempool eviction reasons. I'll review/pocket https://github.com/0xB10C/bitcoin/commit/a06f5c7db6c0e491f150fe5dafb864c15326bb71# for now but if we decide to break format again, I think it's worth inclusion.

instagibbs commented on src/wallet/wallet.cpp:
```diff
@@ -1227,7 +1227,7 @@ void CWallet::blockConnected(const CBlock& block, int height)
     for (size_t index = 0; index < block.vtx.size(); index++) {
         SyncTransaction(block.vtx[index], {CWalletTx::Status::CONFIRMED, height, block_hash, (int)index});
-        transactionRemovedFromMempool(block.vtx[index], MemPoolRemovalReason::BLOCK);
+        transactionRemovedFromMempool(block.vtx[index], MemPoolRemovalReason::BLOCK, 0 /* mempool_sequence */);
```
it's a required chain interface argument, but the wallet doesn't care/have access to about the mempool sequence value.

instagibbs commented:
rebased. No functional changes.

jonatack commented on src/zmq/zmqnotificationinterface.cpp:
```diff
@@ -140,31 +141,53 @@ void CZMQNotificationInterface::UpdatedBlockTip(const CBlockIndex *pindexNew, co
 }
 
-void CZMQNotificationInterface::TransactionAddedToMempool(const CTransactionRef& ptx)
+void CZMQNotificationInterface::TransactionAddedToMempool(const CTransactionRef& ptx,  uint64_t mempool_sequence)
```
pico-nit 68b1e612 from last rebase if you have to retouch again
```suggestion
void CZMQNotificationInterface::TransactionAddedToMempool(const CTransactionRef& ptx, uint64_t mempool_sequence)
```

instagibbs replied:
you're in luck, rebasing again. done.

jonatack reviewed (commented):
Code review re-ACK 877f511d7f1f3c37a86c421e21fc0694c387e69f per `git range-diff c0c409d 4baaa89 877f511`, looked at the new function `TryForEachAndRemoveFailed()` that caused most of the rebase, verified debug build is clean and the changed test `interface_zmq.py` passes.

n-thumann commented:
re-ACK https://github.com/bitcoin/bitcoin/commit/877f511d7f1f3c37a86c421e21fc0694c387e69f

instagibbs commented:
rebased again

laanwj commented:
Code review ACK 759d94e70f6844443106404882c7b105f3a4dba7

I see a potential confusion between sequence numbers of the individual packets and the "mempool sequence" but this is a matter of documentation, I don't have a suggestion for better naming.

instagibbs commented:
> I see a potential confusion between sequence numbers of the individual packets and the "mempool sequence" but this is a matter of documentation, I don't have a suggestion for better naming.

Willing to entertain any renaming if people think of something better!

jonatack commented:
re-ACK 759d94e70f6844443106404882c7b105f3a4dba7

shuckc commented:
Lovely! Very much look forward to using this.
